File: debop4s-timeperiod/src/test/scala/debop4s/timeperiod/tests/timeranges/MinuteRangeTest.scala
package debop4s.timeperiod.tests.timeranges

import debop4s.core.jodatime._
import debop4s.timeperiod._
import debop4s.timeperiod.tests.AbstractTimePeriodTest
import debop4s.timeperiod.timerange.{HourRange, MinuteRange}
import debop4s.timeperiod.utils.Times

/**
 * MinuteRangeTest
 * Created by debop on 2014. 2. 16.
 */
class MinuteRangeTest extends AbstractTimePeriodTest {

    test("init values") {
        val now = Times.now
        val firstMin = Times.trimToSecond(now)
        val secondMin = firstMin + 1.minute

        val minRange = MinuteRange(now, EmptyOffsetTimeCalendar)

        minRange.start.getYear shouldEqual firstMin.getYear
        minRange.start.getMonthOfYear shouldEqual firstMin.getMonthOfYear
        minRange.start.getDayOfMonth shouldEqual firstMin.getDayOfMonth
        minRange.start.getHourOfDay shouldEqual firstMin.getHourOfDay
        minRange.start.getMinuteOfHour shouldEqual firstMin.getMinuteOfHour
        minRange.start.getSecondOfMinute shouldEqual 0
        minRange.start.getMillisOfSecond shouldEqual 0
    }

    test("default calendar") {
        val now = Times.now
        val today = Times.today

        (0 until MinutesPerHour).par.foreach { m =>
            val mr = MinuteRange(today + m.minute)

            mr.year shouldEqual today.getYear
            mr.monthOfYear shouldEqual today.getMonthOfYear
            mr.dayOfMonth shouldEqual today.getDayOfMonth
            mr.hourOfDay shouldEqual today.getHourOfDay
            mr.minuteOfHour shouldEqual m
            mr.start shouldEqual mr.calendar.mapStart(today + m.minute)
            mr.end shouldEqual mr.calendar.mapEnd(today + (m + 1).minute)
        }
    }

    test("constructor") {
        val now = Times.now

        val mr = MinuteRange(now)
        mr.year shouldEqual now.getYear
        mr.monthOfYear shouldEqual now.getMonthOfYear
        mr.dayOfMonth shouldEqual now.getDayOfMonth
        mr.hourOfDay shouldEqual now.getHourOfDay
        mr.minuteOfHour shouldEqual now.getMinuteOfHour

        val mr2 = MinuteRange(now.getYear, now.getMonthOfYear, now.getDayOfMonth, now.getHourOfDay, now.getMinuteOfHour)
        mr2.year shouldEqual now.getYear
        mr2.monthOfYear shouldEqual now.getMonthOfYear
        mr2.dayOfMonth shouldEqual now.getDayOfMonth
        mr2.hourOfDay shouldEqual now.getHourOfDay
        mr2.minuteOfHour shouldEqual now.getMinuteOfHour
    }

    test("add minutes") {
        val mr = MinuteRange()

        mr.previousMinute.minuteOfHour shouldEqual mr.start.plusMinutes(-1).getMinuteOfHour
        mr.nextMinute.minuteOfHour shouldEqual mr.start.plusMinutes(1).getMinuteOfHour

        val mr2 = MinuteRange(EmptyOffsetTimeCalendar)
        mr2.addMinutes(0) shouldEqual mr2

        val prevRange = mr2.previousMinute
        val expectedPrevRange = mr2.addMinutes(-1)
        prevRange.year shouldEqual expectedPrevRange.year
        prevRange.monthOfYear shouldEqual expectedPrevRange.monthOfYear
        prevRange.dayOfMonth shouldEqual expectedPrevRange.dayOfMonth
        prevRange.hourOfDay shouldEqual expectedPrevRange.hourOfDay
        prevRange.minuteOfHour shouldEqual expectedPrevRange.minuteOfHour

        val nextRange = mr2.nextMinute
        val expectedNextRange = mr2.addMinutes(1)
        nextRange.year shouldEqual expectedNextRange.year
        nextRange.monthOfYear shouldEqual expectedNextRange.monthOfYear
        nextRange.dayOfMonth shouldEqual expectedNextRange.dayOfMonth
        nextRange.hourOfDay shouldEqual expectedNextRange.hourOfDay
        nextRange.minuteOfHour shouldEqual expectedNextRange.minuteOfHour

        (-100 to 100).par.foreach { m =>
            val r1 = mr2.addMinutes(m)
            val r2 = mr2.addMinutes(m)
            r1.year shouldEqual r2.year
            r1.monthOfYear shouldEqual r2.monthOfYear
            r1.dayOfMonth shouldEqual r2.dayOfMonth
            r1.hourOfDay shouldEqual r2.hourOfDay
            r1.minuteOfHour shouldEqual r2.minuteOfHour
        }
    }

    test("getMinutes") {
        val hr = HourRange()
        val minutes = hr.minutes

        minutes.size shouldEqual MinutesPerHour

        (0 until MinutesPerHour).par.foreach { i =>
            val m = minutes(i)

            m.start shouldEqual (hr.start + i.minute)
            m.unmappedStart shouldEqual (hr.calendar.unmapStart(hr.start) + i.minute)
            m.end shouldEqual m.calendar.mapEnd(hr.start + (i + 1).minute)
            m.unmappedEnd shouldEqual (hr.start + (i + 1).minute)
        }
    }

}

File: debop4s-timeperiod/src/test/java/debop4s/timeperiod/tests/timeranges/MinuteRangeTest.java
package debop4s.timeperiod.tests.timeranges;

import debop4s.timeperiod.TimeCalendar;
import debop4s.timeperiod.TimeSpec;
import debop4s.timeperiod.tests.TimePeriodTestBase;
import debop4s.timeperiod.timerange.HourRange;
import debop4s.timeperiod.timerange.MinuteRange;
import debop4s.timeperiod.utils.Times;
import lombok.extern.slf4j.Slf4j;
import org.joda.time.DateTime;
import org.junit.Test;

import java.util.List;

import static org.fest.assertions.Assertions.assertThat;


@Slf4j
public class MinuteRangeTest extends TimePeriodTestBase {
    @Test
    public void initValues() {
        DateTime now = Times.now();
        DateTime firstMin = Times.trimToSecond(now);
        DateTime secondMin = firstMin.plusMinutes(1);

        MinuteRange minRange = new MinuteRange(now, TimeCalendar.getEmptyOffset());

        assertThat(minRange.getStart().getYear()).isEqualTo(firstMin.getYear());
        assertThat(minRange.getStart().getMonthOfYear()).isEqualTo(firstMin.getMonthOfYear());
        assertThat(minRange.getStart().getDayOfMonth()).isEqualTo(firstMin.getDayOfMonth());
        assertThat(minRange.getStart().getMinuteOfHour()).isEqualTo(firstMin.getMinuteOfHour());
        assertThat(minRange.getStart().getSecondOfMinute()).isEqualTo(0);
        assertThat(minRange.getStart().getMillisOfSecond()).isEqualTo(0);

        assertThat(minRange.getEnd().getYear()).isEqualTo(secondMin.getYear());
        assertThat(minRange.getEnd().getMonthOfYear()).isEqualTo(secondMin.getMonthOfYear());
        assertThat(minRange.getEnd().getDayOfMonth()).isEqualTo(secondMin.getDayOfMonth());
        assertThat(minRange.getEnd().getMinuteOfHour()).isEqualTo(secondMin.getMinuteOfHour());
        assertThat(minRange.getEnd().getSecondOfMinute()).isEqualTo(0);
        assertThat(minRange.getEnd().getMillisOfSecond()).isEqualTo(0);
    }

    @Test
    public void defaultCalendar() {
        DateTime now = Times.now();
        DateTime today = Times.today();

        for (int m = 0; m < TimeSpec.MinutesPerHour; m++) {
            MinuteRange minRange = new MinuteRange(today.plusMinutes(m));

            assertThat(minRange.getYear()).isEqualTo(today.getYear());
            assertThat(minRange.getMonthOfYear()).isEqualTo(today.getMonthOfYear());
            assertThat(minRange.getDayOfMonth()).isEqualTo(today.getDayOfMonth());
            assertThat(minRange.getHourOfDay()).isEqualTo(today.getHourOfDay());
            assertThat(minRange.getMinuteOfHour()).isEqualTo(m);
            assertThat(minRange.getStart()).isEqualTo(minRange.getTimeCalendar().mapStart(today.plusMinutes(m)));
            assertThat(minRange.getEnd()).isEqualTo(minRange.getTimeCalendar().mapEnd(today.plusMinutes(m + 1)));
        }
    }

    @Test
    public void constructorTest() {
        DateTime now = Times.now();

        MinuteRange minRange = new MinuteRange(now);
        assertThat(minRange.getYear()).isEqualTo(now.getYear());
        assertThat(minRange.getMonthOfYear()).isEqualTo(now.getMonthOfYear());
        assertThat(minRange.getDayOfMonth()).isEqualTo(now.getDayOfMonth());
        assertThat(minRange.getHourOfDay()).isEqualTo(now.getHourOfDay());
        assertThat(minRange.getMinuteOfHour()).isEqualTo(now.getMinuteOfHour());

        minRange = new MinuteRange(now.getYear(), now.getMonthOfYear(), now.getDayOfMonth(), now.getHourOfDay(), now.getMinuteOfHour());
        assertThat(minRange.getYear()).isEqualTo(now.getYear());
        assertThat(minRange.getMonthOfYear()).isEqualTo(now.getMonthOfYear());
        assertThat(minRange.getDayOfMonth()).isEqualTo(now.getDayOfMonth());
        assertThat(minRange.getHourOfDay()).isEqualTo(now.getHourOfDay());
        assertThat(minRange.getMinuteOfHour()).isEqualTo(now.getMinuteOfHour());
    }

    @Test
    public void addHourTest() {
        MinuteRange minRange = new MinuteRange();

        assertThat(minRange.previousMinute().getMinuteOfHour()).isEqualTo(minRange.getStart().plusMinutes(-1).getMinuteOfHour());
        assertThat(minRange.nextMinute().getMinuteOfHour()).isEqualTo(minRange.getStart().plusMinutes(1).getMinuteOfHour());

        minRange = new MinuteRange(TimeCalendar.getEmptyOffset());

        assertThat(minRange.addMinutes(0)).isEqualTo(minRange);

        MinuteRange prevRange = minRange.previousMinute();
        assertThat(minRange.addMinutes(-1).getYear()).isEqualTo(prevRange.getYear());
        assertThat(minRange.addMinutes(-1).getMonthOfYear()).isEqualTo(prevRange.getMonthOfYear());
        assertThat(minRange.addMinutes(-1).getDayOfMonth()).isEqualTo(prevRange.getDayOfMonth());
        assertThat(minRange.addMinutes(-1).getHourOfDay()).isEqualTo(prevRange.getHourOfDay());
        assertThat(minRange.addMinutes(-1).getMinuteOfHour()).isEqualTo(prevRange.getMinuteOfHour());

        MinuteRange nextRange = minRange.nextMinute();
        assertThat(minRange.addMinutes(1).getYear()).isEqualTo(nextRange.getYear());
        assertThat(minRange.addMinutes(1).getMonthOfYear()).isEqualTo(nextRange.getMonthOfYear());
        assertThat(minRange.addMinutes(1).getDayOfMonth()).isEqualTo(nextRange.getDayOfMonth());
        assertThat(minRange.addMinutes(1).getHourOfDay()).isEqualTo(nextRange.getHourOfDay());
        assertThat(minRange.addMinutes(1).getMinuteOfHour()).isEqualTo(nextRange.getMinuteOfHour());

        for (int m = -100; m < 100; m += 5) {
            MinuteRange range = minRange.addMinutes(m);
            MinuteRange range2 = minRange.addMinutes(m);
            assertThat(range.getYear()).isEqualTo(range2.getYear());
            assertThat(range.getMonthOfYear()).isEqualTo(range2.getMonthOfYear());
            assertThat(range.getDayOfMonth()).isEqualTo(range2.getDayOfMonth());
            assertThat(range.getHourOfDay()).isEqualTo(range2.getHourOfDay());
            assertThat(range.getMinuteOfHour()).isEqualTo(range2.getMinuteOfHour());
        }
    }

    @Test
    public void getMinutesTest() {
        HourRange hourRange = new HourRange();
        List<MinuteRange> minutes = hourRange.minutes();

        assertThat(minutes.size()).isEqualTo(TimeSpec.MinutesPerHour);

        for (int i = 0; i < TimeSpec.MinutesPerHour; i++) {
            MinuteRange minute = minutes.get(i);

            assertThat(minute.getStart()).isEqualTo(hourRange.getStart().plusMinutes(i));
            assertThat(minute.getUnmappedStart()).isEqualTo(hourRange.getStart().plusMinutes(i));

            assertThat(minute.getEnd()).isEqualTo(minute.getTimeCalendar().mapEnd(hourRange.getStart().plusMinutes(i + 1)));
            assertThat(minute.getUnmappedEnd()).isEqualTo(hourRange.getStart().plusMinutes(i + 1));
        }
    }
}
